File: src/main/python/StringMatchingInAnArray_1408.py
from typing import List


class Solution:
    def stringMatching(self, words: List[str]) -> List[str]:
        res = []
        for i, word in enumerate(words):
            if word in ' '.join(words[:i]) or word in ' '.join(words[i+1:]):
                # print(word, ' '.join(words[:i]), ' '.join(words[i+1:]))
                res.append(word)

        return res

File: src/main/scala/StringMatchingInAnArray_1408.scala
object StringMatchingInAnArray_1408 {
  def stringMatching1(words: Array[String]): List[String] = {
    var res = List[String]()
    for(i <- words.indices){
      var found = false
      for (j <- words.indices if !found){
        if(i != j && words(j).contains(words(i))){
          res = res :+ words(i)
          found = true
        }
      }
    }
    res
  }

  // FP
  def stringMatching(words: Array[String]): List[String] = {
    words.indices.flatMap(i =>
      words.indices.collect{
        case j if i != j && words(j).contains(words(i)) => words(i)
      }
    ).distinct.toList
  }
}